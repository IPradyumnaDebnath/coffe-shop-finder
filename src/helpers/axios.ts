import axios from "axios";
import MockAdaptor from "axios-mock-adapter";

export const api = axios.create({
  baseURL: "https://api.foursquare.com/v3",
  headers: {
    //Should be kept in .env for local and environment variables of CI/CD Pipeline
    Authorization: "fsq37bPXSaW8IljjjY49FQFnQ1PcRnA2dcH0OAcnqyvB7GA=",
  },
});

export const mockApi = new MockAdaptor(axios);
