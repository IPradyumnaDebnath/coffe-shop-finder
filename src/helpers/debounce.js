/**
* custom hook to get debounced value
* @params
* func: callback to be debounces
* wait: delay in ms
* @returns debounced function
*/
export default function debounce(func, wait) {
  let timeout;
  return function () {
    const context = this;
    const args = arguments;
    const later = function () {
      timeout = null;
      func.apply(context, args);
    };
    clearTimeout(timeout);
    timeout = setTimeout(later, wait);
  };
}
