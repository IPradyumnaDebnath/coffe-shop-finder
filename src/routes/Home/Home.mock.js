export const Places = {
  results: [
    {
      fsq_id: "52004fe52fc61d655eccc172",
      categories: [
        {
          id: 13034,
          name: "Café",
          icon: {
            prefix: "https://ss3.4sqi.net/img/categories_v2/food/cafe_",
            suffix: ".png",
          },
        },
      ],
      chains: [],
      distance: 7766,
      geocodes: {
        main: {
          latitude: 39.115978,
          longitude: -77.564289,
        },
      },
      link: "/v3/places/52004fe52fc61d655eccc172",
      location: {
        address: "17 N King St",
        census_block: "511076105052005",
        country: "US",
        cross_street: "Market Street",
        dma: "Washington, Dc-Hagrstwn",
        formatted_address: "17 N King St (Market Street), Leesburg, VA 20176",
        locality: "Leesburg",
        postcode: "20176",
        region: "VA",
      },
      name: "Shoe’s Cup and Cork",
      related_places: {
        children: [
          {
            fsq_id: "5ce16d0a9be522002cb22162",
            name: "The Secret Garden",
          },
        ],
      },
      timezone: "America/New_York",
    },
    {
      fsq_id: "4b4e151cf964a520e9e026e3",
      categories: [
        {
          id: 13065,
          name: "Restaurant",
          icon: {
            prefix: "https://ss3.4sqi.net/img/categories_v2/food/default_",
            suffix: ".png",
          },
        },
      ],
      chains: [],
      distance: 7696,
      geocodes: {
        main: {
          latitude: 39.112704,
          longitude: -77.56297,
        },
        roof: {
          latitude: 39.112704,
          longitude: -77.56297,
        },
      },
      link: "/v3/places/4b4e151cf964a520e9e026e3",
      location: {
        address: "203 Harrison St SE",
        address_extended: "Ste J",
        census_block: "511076105052010",
        country: "US",
        cross_street: "",
        dma: "Washington, Dc-Hagrstwn",
        formatted_address: "203 Harrison St SE, Leesburg, VA 20175",
        locality: "Leesburg",
        postcode: "20175",
        region: "VA",
      },
      name: "South Street Under",
      related_places: {},
      timezone: "America/New_York",
    },
    {
      fsq_id: "4fceb17ce4b01e8cab29f632",
      categories: [
        {
          id: 13035,
          name: "Coffee Shop",
          icon: {
            prefix: "https://ss3.4sqi.net/img/categories_v2/food/coffeeshop_",
            suffix: ".png",
          },
        },
      ],
      chains: [
        {
          id: "ab4c54c0-d68a-012e-5619-003048cad9da",
          name: "Starbucks",
        },
      ],
      distance: 7110,
      geocodes: {
        main: {
          latitude: 39.136494,
          longitude: -77.693046,
        },
        roof: {
          latitude: 39.136494,
          longitude: -77.693046,
        },
      },
      link: "/v3/places/4fceb17ce4b01e8cab29f632",
      location: {
        address: "140A Purcellville Gateway Dr",
        census_block: "511076107021019",
        country: "US",
        cross_street: "at Harris Teeter",
        dma: "Washington, Dc-Hagrstwn",
        formatted_address:
          "140A Purcellville Gateway Dr (at Harris Teeter), Purcellville, VA 20132",
        locality: "Purcellville",
        postcode: "20132",
        region: "VA",
      },
      name: "Starbucks",
      related_places: {},
      timezone: "America/New_York",
    },
    {
      fsq_id: "58a0bed492789f6a81b9b3c8",
      categories: [
        {
          id: 13025,
          name: "Wine Bar",
          icon: {
            prefix: "https://ss3.4sqi.net/img/categories_v2/food/winery_",
            suffix: ".png",
          },
        },
        {
          id: 13034,
          name: "Café",
          icon: {
            prefix: "https://ss3.4sqi.net/img/categories_v2/food/cafe_",
            suffix: ".png",
          },
        },
        {
          id: 13041,
          name: "Creperie",
          icon: {
            prefix: "https://ss3.4sqi.net/img/categories_v2/food/creperie_",
            suffix: ".png",
          },
        },
      ],
      chains: [],
      distance: 7644,
      geocodes: {
        main: {
          latitude: 39.137334,
          longitude: -77.703822,
        },
        roof: {
          latitude: 39.137334,
          longitude: -77.703822,
        },
      },
      link: "/v3/places/58a0bed492789f6a81b9b3c8",
      location: {
        address: "713 E Main St",
        census_block: "511076107021011",
        country: "US",
        cross_street: "",
        dma: "Washington, Dc-Hagrstwn",
        formatted_address: "713 E Main St, Purcellville, VA 20132",
        locality: "Purcellville",
        postcode: "20132",
        region: "VA",
      },
      name: "La Petite LouLou Creperie",
      related_places: {},
      timezone: "America/New_York",
    },
    {
      fsq_id: "4c93911758d4b60cb05c2029",
      categories: [
        {
          id: 13035,
          name: "Coffee Shop",
          icon: {
            prefix: "https://ss3.4sqi.net/img/categories_v2/food/coffeeshop_",
            suffix: ".png",
          },
        },
        {
          id: 13065,
          name: "Restaurant",
          icon: {
            prefix: "https://ss3.4sqi.net/img/categories_v2/food/default_",
            suffix: ".png",
          },
        },
      ],
      chains: [],
      distance: 7576,
      geocodes: {
        main: {
          latitude: 39.113795,
          longitude: -77.565169,
        },
        roof: {
          latitude: 39.113795,
          longitude: -77.565169,
        },
      },
      link: "/v3/places/4c93911758d4b60cb05c2029",
      location: {
        address: "110 S King St",
        census_block: "511076105052009",
        country: "US",
        dma: "Washington, Dc-Hagrstwn",
        formatted_address: "110 S King St, Leesburg, VA 20175",
        locality: "Leesburg",
        postcode: "20175",
        region: "VA",
      },
      name: "King Street Coffee",
      related_places: {},
      timezone: "America/New_York",
    },
    {
      fsq_id: "50c63738e4b030063396558e",
      categories: [
        {
          id: 13034,
          name: "Café",
          icon: {
            prefix: "https://ss3.4sqi.net/img/categories_v2/food/cafe_",
            suffix: ".png",
          },
        },
        {
          id: 13035,
          name: "Coffee Shop",
          icon: {
            prefix: "https://ss3.4sqi.net/img/categories_v2/food/coffeeshop_",
            suffix: ".png",
          },
        },
      ],
      chains: [],
      distance: 9542,
      geocodes: {
        main: {
          latitude: 39.106807,
          longitude: -77.537868,
        },
        roof: {
          latitude: 39.106807,
          longitude: -77.537868,
        },
      },
      link: "/v3/places/50c63738e4b030063396558e",
      location: {
        address: "241 Fort Evans Rd NE",
        census_block: "511076105061000",
        country: "US",
        cross_street: "",
        dma: "Washington, Dc-Hagrstwn",
        formatted_address: "241 Fort Evans Rd NE, Leesburg, VA 20176",
        locality: "Leesburg",
        postcode: "20176",
        region: "VA",
      },
      name: "McCawley Coffee",
      related_places: {},
      timezone: "America/New_York",
    },
    {
      fsq_id: "4b6c276cf964a52082262ce3",
      categories: [
        {
          id: 13035,
          name: "Coffee Shop",
          icon: {
            prefix: "https://ss3.4sqi.net/img/categories_v2/food/coffeeshop_",
            suffix: ".png",
          },
        },
      ],
      chains: [
        {
          id: "ab4c54c0-d68a-012e-5619-003048cad9da",
          name: "Starbucks",
        },
      ],
      distance: 8417,
      geocodes: {
        main: {
          latitude: 39.108351,
          longitude: -77.552227,
        },
        roof: {
          latitude: 39.108351,
          longitude: -77.552227,
        },
      },
      link: "/v3/places/4b6c276cf964a52082262ce3",
      location: {
        address: "526 E Market St",
        address_extended: "Ste D",
        census_block: "511076105051002",
        country: "US",
        cross_street: "at Bellwood Commons",
        dma: "Washington, Dc-Hagrstwn",
        formatted_address:
          "526 E Market St (at Bellwood Commons), Leesburg, VA 20176",
        locality: "Leesburg",
        postcode: "20176",
        region: "VA",
      },
      name: "Starbucks",
      related_places: {},
      timezone: "America/New_York",
    },
    {
      fsq_id: "4b526032f964a5202f7a27e3",
      categories: [
        {
          id: 13001,
          name: "Bagel Shop",
          icon: {
            prefix: "https://ss3.4sqi.net/img/categories_v2/food/bagels_",
            suffix: ".png",
          },
        },
      ],
      chains: [
        {
          id: "2cb519f8-883c-4263-860a-cd83325fbb97",
          name: "Dunkin'",
        },
      ],
      distance: 8384,
      geocodes: {
        main: {
          latitude: 39.108853,
          longitude: -77.552708,
        },
        roof: {
          latitude: 39.108853,
          longitude: -77.552708,
        },
      },
      link: "/v3/places/4b526032f964a5202f7a27e3",
      location: {
        address: "521 E Market St",
        address_extended: "Ste J",
        census_block: "511076105051002",
        country: "US",
        dma: "Washington, Dc-Hagrstwn",
        formatted_address: "521 E Market St, Leesburg, VA 20176",
        locality: "Leesburg",
        postcode: "20176",
        region: "VA",
      },
      name: "Dunkin'",
      related_places: {},
      timezone: "America/New_York",
    },
    {
      fsq_id: "4bba04a57421a5937f44c340",
      categories: [
        {
          id: 13059,
          name: "Juice Bar",
          icon: {
            prefix: "https://ss3.4sqi.net/img/categories_v2/food/juicebar_",
            suffix: ".png",
          },
        },
      ],
      chains: [
        {
          id: "2cd1878e-99c9-4d55-afbf-da408239814e",
          name: "Tropical Smoothie Cafe",
        },
      ],
      distance: 7805,
      geocodes: {
        main: {
          latitude: 39.137463,
          longitude: -77.706972,
        },
        roof: {
          latitude: 39.137463,
          longitude: -77.706972,
        },
      },
      link: "/v3/places/4bba04a57421a5937f44c340",
      location: {
        address: "609 E Main St",
        address_extended: "Ste BB",
        census_block: "511076107021006",
        country: "US",
        cross_street: "",
        dma: "Washington, Dc-Hagrstwn",
        formatted_address: "609 E Main St, Purcellville, VA 20132",
        locality: "Purcellville",
        postcode: "20132",
        region: "VA",
      },
      name: "Tropical Smoothie Cafe",
      related_places: {},
      timezone: "America/New_York",
    },
    {
      fsq_id: "4be18bf48e2962b55e589b0a",
      categories: [
        {
          id: 13035,
          name: "Coffee Shop",
          icon: {
            prefix: "https://ss3.4sqi.net/img/categories_v2/food/coffeeshop_",
            suffix: ".png",
          },
        },
      ],
      chains: [
        {
          id: "ab4c54c0-d68a-012e-5619-003048cad9da",
          name: "Starbucks",
        },
      ],
      distance: 10070,
      geocodes: {
        main: {
          latitude: 39.11285,
          longitude: -77.535175,
        },
      },
      link: "/v3/places/4be18bf48e2962b55e589b0a",
      location: {
        address: "1011 Edwards Ferry Rd NE",
        census_block: "511076105033006",
        country: "US",
        cross_street: "at Battlefield Ctr",
        dma: "Washington, Dc-Hagrstwn",
        formatted_address:
          "1011 Edwards Ferry Rd NE (at Battlefield Ctr), Leesburg, VA 20176",
        locality: "Leesburg",
        postcode: "20176",
        region: "VA",
      },
      name: "Starbucks",
      related_places: {},
      timezone: "America/New_York",
    },
  ]
};
