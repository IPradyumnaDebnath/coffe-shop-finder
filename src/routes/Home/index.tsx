import { useEffect, useState, useCallback } from "react";

import { api } from "helpers/axios";
import debounce from "helpers/debounce";
import Search from "components/SearchInput";
import CoffeeShops from "components/CoffeeShops";
import { Shop } from "components/CoffeeShop";

type CoffeeShopState = {
  persisted: [Shop] | [];
  current: [Shop] | [];
  loading: boolean;
  next?: string;
};

const Home = () => {
  const [location, setLocation] = useState("");
  const [coffeeShops, setCoffeeShops] = useState<CoffeeShopState>({
    persisted: [],
    current: [],
    loading: true,
    next: "",
  });
  const fetchAndUpdateCoffeeShops = useCallback(
    debounce((location: string, cursor: string | undefined) => {
      setCoffeeShops((prevShops) => ({ ...prevShops, loading: true }));
      api
        .get(`/places/search`, {
          params: {
            query: location,
            categories: "11126,13032,13033,13034,13035,13036,13063",
            limit: "3",
            sort: "DISTANCE",
            cursor,
          },
        })
        .then((resp) => {
          const nextPageLinkParams = new URLSearchParams(
            resp.headers["link"] ?? ""
          );
          setCoffeeShops((prev) => ({
            ...prev,
            current: resp.data.results,
            next: nextPageLinkParams.get("cursor") || "",
          }));
        })
        .catch((err) => console.error(err))
        .finally(() => {
          setCoffeeShops((curShops) => ({ ...curShops, loading: false }));
        });
    }, 500),
    []
  );
  const handleQueryUpdate = (loc: string) => {
    fetchAndUpdateCoffeeShops(loc);
    setLocation(loc);
  };

  useEffect(() => {
    fetchAndUpdateCoffeeShops();
    window.addEventListener("scroll", handlePageScroll);
    return () => window.removeEventListener("scroll", handlePageScroll);
  }, []);

  function handlePageScroll() {
    if (
      window.innerHeight + document.documentElement.scrollTop !==
      document.documentElement.offsetHeight
    )
      return;

    fetchAndUpdateCoffeeShops(location, coffeeShops.next);
  }

  return (
    <div className="mt-5 space-y-6">
      <header>
        <div className="px-4 py-2 m-auto w-fit text-4xl shadow-xl bg-slate-200 font-bold uppercase">
          Coffee Corners
        </div>
      </header>
      <Search location={location} onChange={handleQueryUpdate} />
      <CoffeeShops
        shopList={coffeeShops.current}
        loading={coffeeShops?.loading }
      />
    </div>
  );
};

export default Home;
