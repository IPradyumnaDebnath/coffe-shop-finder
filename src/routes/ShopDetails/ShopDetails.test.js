import renderer from "react-test-renderer";

import ShopDetailsPage from "./index";
import { mockApi } from "helpers/axios";
import { Places } from "../Home/Home.mock";

it("renders correctly", () => {
  mockApi.onGet("/places/search").reply(200, { ...Places.results[0] });
  const tree = renderer.create(<ShopDetailsPage />).toJSON();
  expect(tree).toMatchSnapshot();
});
