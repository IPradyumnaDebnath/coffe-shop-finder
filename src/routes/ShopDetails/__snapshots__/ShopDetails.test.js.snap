// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="flex justify-center items-center"
>
  <div>
    Loading...
  </div>
  <div
    className="spinner-border animate-spin inline-block w-8 h-8 border-4 rounded-full"
    role="status"
  />
</div>
`;
